defaults/validate.go

```diff
@@ -0,0 +1,264 @@
+package defaults
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+var (
+	emailRegex = regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
+	rangeRegex = regexp.MustCompile(`range:(\[|\()(\S*)\,(\S*)(\]|\))`)
+	oneofRegex = regexp.MustCompile(`oneof:(\{)(.*)(\})`)
+)
+
+// Validate validates each field of the value
+func Validate(value interface{}) error {
+	v := reflect.Indirect(reflect.ValueOf(value))
+	t := v.Type()
+
+	m := reflect.ValueOf(value).MethodByName("IsValid")
+	if m.IsValid() {
+		e := m.Call([]reflect.Value{})
+		err, ok := e[0].Interface().(error)
+		if ok && err != nil {
+			return err
+		}
+	}
+
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		switch field.Type().Kind() {
+		case reflect.Struct:
+			dv := field.Interface()
+			if err := Validate(dv); err != nil {
+				return err
+			}
+		case reflect.Slice, reflect.Map:
+			dv := reflect.ValueOf(field.Interface())
+			for j := 0; j < dv.Len(); j++ {
+				if err := Validate(dv.Index(j).Interface()); err != nil {
+					return err
+				}
+			}
+		case reflect.Bool, reflect.Int, reflect.Float64, reflect.String:
+			tag, ok := t.Field(i).Tag.Lookup("validate")
+			if !ok {
+				continue
+			}
+			if err := validate(tag, t.Field(i).Name, v, v.Field(i)); err != nil {
+				return err
+			}
+		case reflect.Chan:
+			return nil
+		default:
+			return fmt.Errorf("unimplemented struct field type: %s", t.Field(i).Name)
+		}
+	}
+	return nil
+}
+
+func validate(validation, fieldName string, p, v reflect.Value) error {
+	switch validation {
+	case "url":
+		s := v.String()
+		_, err := url.ParseRequestURI(s)
+		if err != nil {
+			return err
+		}
+	case "email":
+		s := v.String()
+		if !emailRegex.MatchString(s) {
+			return fmt.Errorf("%s is not a valid e-mail address", s)
+		}
+	case "notempty":
+		switch v.Type().Kind() {
+		case reflect.String:
+			s := v.String()
+			if s == "" {
+				return fmt.Errorf("%s is empty", fieldName)
+			}
+		}
+	case "alpha":
+		s := v.String()
+		if s == "" || !isAlphanumeric(s) {
+			return fmt.Errorf("%s is not alphanumeric", fieldName)
+		}
+	case "file":
```

Comment:
I am not 100% convinced by this. What about files that we write to (e.g. log files) which still could benefit from some basic parsing. 

I think we should limit the `file` validation to make sure the string is a valid filepath and add a new validation rule for the file existence which will be used only on input files.